Encapsulate part of RAFT state

jabolina opened the pull request:
Fixes #218. It took some time to get everything. I was delving into the resources to understand what happens after the term is updated. Everything should be covered now. On term update:

* Reset the term's leader and vote;
* Step down already exists, but it seems the role changes weren't applied in all updates;

I was worried that the leader needed to handle responses with higher terms differently, but it seemed unnecessary. Added a small test that ensures the leader is reset when the term increases. This leads to a change in role, too. The commit includes:

* Encapsulate the current term, the leader, and vote together. These values are updated atomically. This maintains some of the leader election invariants.
* Advancing the term reset the leader and vote to null.
* Restrict the leader and vote updates within the same term.
* Updates the test to the way terms and leader is updated now.

belaban commented:
Hey Jose
this PR makes `SyncElectionTestWithRestriction` fail: `Changing leader S1 to S5 illegally`. Can you take a look?
Also, IMO, `RAFT.setLeaderAndTerm()` is *not* atomic, as it uses 2 *separate* into `RaftState` to set the term and leader.
Cheers,

jabolina commented:
Thanks for the review, Bela! I updated the test name from `SyncElectionTestWithRestriction` -> `SyncElectionWithRestrictionTest` as it seems it was not executing with `mvn clean test`. Also, I addressed the `RAFT.setLeaderAndTerm()` to an atomic update.

belaban commented:
Looks good now; also, the aforementioned test passes now
